src/mscorlib/shared/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs

```diff
@@ -203,6 +221,92 @@ public override void Return(T[] array, bool clearArray = false)
+
+            foreach (PerCoreLockedStacks bucket in _buckets)
+            {
+                bucket?.Trim((uint)milliseconds, Id, pressure, _bucketArraySizes);
+            }
+
+            if (pressure == MemoryPressure.High)
+            {
+                // Under high pressure, release all thread locals
+                foreach (KeyValuePair<T[][], object> tlsBuckets in s_AllTlsBuckets)
+                {
+                    T[][] buckets = tlsBuckets.Key;
+                    for (int i = 0; i < NumBuckets; i++)
+                    {
+                        T[] buffer = buckets[i];
+                        buckets[i] = null;
+
+                        if (log.IsEnabled() && buffer != null)
+                        {
+                            log.BufferTrimmed(buffer.GetHashCode(), buffer.Length, Id);
+                        }
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// This is the static function that is called from the gen2 GC callback.
+        /// The input object is the instance we want the callback on.
+        /// </summary>
+        /// <remarks>
+        /// The reason that we make this function static and take the instance as a parameter is that
+        /// we would otherwise root the instance to the Gen2GcCallback object, leaking the instance even when
+        /// the application no longer needs it.
+        /// </remarks>
+        private static bool Gen2GcCallbackFunc(object target)
+        {
+            return ((TlsOverPerCoreLockedStacksArrayPool<T>)(target)).Trim();
+        }
+
+        private enum MemoryPressure
+        {
+            Low,
+            Medium,
+            High
+        }
+
+        private static MemoryPressure GetMemoryPressure()
+        {
+            const double HighPressureThreshold = .90;                       // Percent of GC memory pressure threshold we consider "high"
+            const double MediumPressureThreshold = .70;                     // Percent of GC memory pressure threshold we consider "medium"
+
+            GC.GetMemoryInfo(out uint threshold, out _, out uint lastLoad, out _, out _);
+            if (lastLoad >= threshold * HighPressureThreshold)
+            {
+                return MemoryPressure.High;
+            }
+            else if (lastLoad >= threshold * MediumPressureThreshold)
+            {
+                return MemoryPressure.Medium;
+            }
+            return MemoryPressure.Low;
+        }
+
+        private static bool GetTrimBuffers()
+        {
+            // Environment uses ArrayPool, so we have to hit the API directly.
+#if !CORECLR
+            // P/Invokes are different for CoreCLR/RT- for RT we'll not allow
+            // enabling/disabling for now.
+            return true;
+#else
+            return CLRConfig.GetBoolValueWithFallbacks("System.Buffers.TrimBuffers", "DOTNET_SYSTEM_BUFFERS_TRIMBUFFERS", defaultValue: true);
```

Comment:
Could we use a name that's specific to ArrayPool, or even to ArrayPool.Shared, rather than for the general System.Buffers namespace?

Reply:
Sure